state-chain/runtime/src/event_encoder.rs

```diff
@@ -0,0 +1,270 @@
+use codec::{Decode, Encode};
+use scale_info::{prelude::string::String, PortableRegistry};
+// use scale_value::{Value, ValueDef};
+use serde::{Deserialize, Serialize};
+
+use crate::RuntimeEvent;
+
+#[derive(Encode, Decode, Clone, Copy, Eq, PartialEq, Serialize, Deserialize, Debug)]
+pub enum EventDecoderError {
+	FailedToDecodeFromScaleBytes,
+	FailedToDecodeFromString,
+	FailedToConvertIntoJson,
+}
+
+/// A struct that provides interface for decoding RuntimeEvents.
+pub struct RuntimeEventDecoder {
+	registry: PortableRegistry,
+	type_id: u32,
+}
+
+impl RuntimeEventDecoder {
+	/// Creates and returns an instance of a PortableRegistry, used for decoding Runtime Events.
+	pub fn new() -> Self {
+		// We can get the 'portable' type info using scale_info.
+		let meta = scale_info::MetaType::new::<RuntimeEvent>();
+		let mut registry = scale_info::Registry::new();
+		let id = registry.register_type(&meta).id;
+
+		Self { registry: PortableRegistry::from(registry), type_id: id }
+	}
+
+	pub fn decode_event_to_string(&self, data: Vec<u8>) -> Result<String, EventDecoderError> {
+		scale_value::scale::decode_as_type(&mut &*data, &self.type_id, &self.registry)
+			.map_err(|_| EventDecoderError::FailedToDecodeFromScaleBytes)
+			.map(|value| value.to_string())
+	}
+
+	pub fn decode_event_to_string_json(&self, data: Vec<u8>) -> Result<String, EventDecoderError> {
+		let value = scale_value::scale::decode_as_type(&mut &*data, &self.type_id, &self.registry)
+			.map_err(|_| EventDecoderError::FailedToDecodeFromScaleBytes)?;
+		value
+			.serialize(scale_value::serde::ValueSerializer)
+			.map_err(|_| EventDecoderError::FailedToConvertIntoJson)
+			.map(|value| value.to_string())
```

Comment:
I think here you need to use `serde_json::to_string` somehow. Or maybe use [scale_value::from_value](https://docs.rs/scale-value/latest/scale_value/serde/fn.from_value.html) to convert a scale Value to a serde_json::Value. Something like:

```rust
// decode_as_type will give us a scale_value::Value that represents something
// of type type_id (ie. our RuntimeEvent)
scale_value::scale::decode_as_type(&mut &*data, &self.type_id, &self.registry)
	// We can try to convert the scale_value::Value to any other type that implements
	// Deserialize. I *think* we might be able to convert to a serde_json::Value:
	.and_then(scale_value::serde::from_value::<_, _, serde_json::Value>)
	// If we have a serde_json::Value, we can convert to valid json:
	.and_then(serde_json::to_string)
```

In any case, scale_value::to_string will not give valid json - this would explain the strange almost-json like syntax (like the `(( 1, 2, 3 ))` arrays).